tools/linker/MonoTouch.Tuner/MonoTouchTypeMap.cs

```diff
@@ -15,52 +15,128 @@
 using Mono.Linker.Steps;
 using Mono.Tuner;
 
+using Xamarin.Bundler;
 using Xamarin.Linker;
 using Xamarin.Tuner;
 
 namespace MonoTouch.Tuner {
 
 	public class MonoTouchTypeMapStep : TypeMapStep {
 		HashSet<TypeDefinition> cached_isnsobject = new HashSet<TypeDefinition> ();
-		HashSet<TypeDefinition> needs_isdirectbinding_check = new HashSet<TypeDefinition> ();
-		HashSet<MethodDefinition> generated_code = new HashSet<MethodDefinition> ();
+		Dictionary<TypeDefinition, bool?> isdirectbinding_value = new Dictionary<TypeDefinition, bool?> ();
+
+		bool dynamic_registration_support_required;
 
 		DerivedLinkContext LinkContext {
 			get {
 				return (DerivedLinkContext) base.Context;
 			}
 		}
 
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			if (!LinkContext.App.Optimizations.RemoveDynamicRegistrar.HasValue)
+				dynamic_registration_support_required |= RequiresDynamicRegistrar (assembly);
+
+			base.ProcessAssembly (assembly);
+		}
+
+		// If certain conditions are met, we can optimize away the code for the dynamic registrar..
+		bool RequiresDynamicRegistrar (AssemblyDefinition assembly)
+		{
+#if MONOMAC && !XAMCORE_2_0
+			// Disable removing the dynamic registrar for XM/Classic to simplify the code a little bit.
```

Comment:
We're not rebuilding XamMac.dll so can't this die when you merge master?

Reply:
The condition is wrong, this is in mmp, which means the XAMCORE_2_0 condition is not taken into account (the condition should be `if (Driver.IsClassic) return true;` or some such).

I've only tested XI/mtouch, but the intention is to support everything that makes even a little bit sense in XM/mmp as well, so I'll fix this.